### DISCUSSION TITLE
Proposed rule: Use delegation of mutable state rather than variable state value

### DISCUSSION THREAD

**Discussion by VirtualParticle** [slackhq/compose-lints#376] [Ideas]
Storing the actual state object in a variable like this:
```kotlin
var x = remember { mutableStateOf(0) }
```
should be avoided in favor of
```kotlin
var x by remember { mutableStateOf(0) }
```
In the first example, there's no reason that the field `x` should actually be mutable since it's the `value` property within it that should be changed. Even in this example:
```kotlin
val x = remember { mutableStateOf(0) }
```
where `x` is properly immutable, it's still probably preferred that the delegation syntax be used just because it's cleaner. I and the official compose docs seem to mostly prefer the delegation syntax, but not everyone might agree.

**Comment by ZacSweers** [comment]
I'm not sure we could make a strongly defined enough lint to make this practical. For example, it's legal to write `val (x, setX) = remember { mutableStateOf(0) }` to capture a live value and setter

**Reply by VirtualParticle** [reply]
Mainly what I'm targeting here is the first example, with `x` being mutable. I think that syntax should be discouraged because it's sort of misleading where the mutability actually is. I agree that an issue with this would be picking what syntax of the three valid ones to recommend:

- `val x = remember { mtuableStateOf(0) }`
- `var x by remember { mutableStateOf(0) }`
- `val (x, setX) = remember { mutableStateOf(0) }`

In my opinion, the first one should pretty much just be avoided and the second should be used unless there's some other particular reason why the third makes sense. That being said, they're all perfectly valid and used in the compose docs so perhaps not worthy of being linted.

The way I'd implement the rule would be to just call out usages of `var x = remember...` (which doesn't really have any valid use cases, as far as I can tell) and tell the user to use one of those three options.

**Reply by ZacSweers** [reply]
I think starting with a soft warning lint that warns on that situation while respecting the other two sounds reasonable if you're interested in trying a PR. I'm less sure about lint's available APIs to unpack destructuring and delegation though, there might be limitations
